source/extensions/filters/listener/connect_handler/connect_handler.cc

```diff
@@ -0,0 +1,162 @@
+
+#include "envoy/event/dispatcher.h"
+#include "envoy/network/listen_socket.h"
+#include "envoy/stats/scope.h"
+
+#include "source/common/api/os_sys_calls_impl.h"
+#include "source/common/common/assert.h"
+#include "source/common/common/macros.h"
+#include "source/common/http/headers.h"
+#include "source/common/http/utility.h"
+
+#include "absl/strings/match.h"
+#include "absl/strings/str_split.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace ListenerFilters {
+namespace ConnectHandler {
+
+Config::Config(Stats::Scope& scope)
+    : stats_{ALL_CONNECT_HANDLER_STATS(POOL_COUNTER_PREFIX(scope, "connect_handler."))} {}
+
+Filter::Filter(const ConfigSharedPtr config) : config_(config) {
+  http_parser_init(&parser_, HTTP_REQUEST);
+}
+
+http_parser_settings Filter::settings_{
+    nullptr, nullptr, nullptr, nullptr, nullptr, nullptr, nullptr, nullptr, nullptr, nullptr,
+};
+
+thread_local uint8_t Filter::buf_[Config::MAX_INSPECT_SIZE];
+static absl::string_view response_body = " 200 Connection Established\r\n\r\n";
+
+Network::FilterStatus Filter::onAccept(Network::ListenerFilterCallbacks& cb) {
+  ENVOY_LOG(debug, "connect handler: new connection accepted");
+
+  const Network::ConnectionSocket& socket = cb.socket();
+
+  const absl::string_view transport_protocol = socket.detectedTransportProtocol();
+  if (!transport_protocol.empty() && transport_protocol != "raw_buffer") {
+    ENVOY_LOG(trace, "connect handler: cannot inspect http protocol with transport socket {}",
+              transport_protocol);
+    return Network::FilterStatus::Continue;
+  }
+
+  cb_ = &cb;
+  const ParseState parse_state = onRead();
+  switch (parse_state) {
+  case ParseState::Error:
+    // As per discussion in https://github.com/envoyproxy/envoy/issues/7864
+    // we don't add new enum in FilterStatus so we have to signal the caller
+    // the new condition.
+    cb.socket().close();
+    return Network::FilterStatus::StopIteration;
+  case ParseState::Done:
+    return Network::FilterStatus::Continue;
+  case ParseState::Continue:
+    // do nothing but create the event
+    cb.socket().ioHandle().initializeFileEvent(
+        cb.dispatcher(),
+        [this](uint32_t events) {
+          ENVOY_LOG(trace, "connect handler event: {}", events);
+
+          const ParseState parse_state = onRead();
+          switch (parse_state) {
+          case ParseState::Error:
+            cb_->socket().ioHandle().resetFileEvents();
+            cb_->continueFilterChain(false);
+            break;
+          case ParseState::Done:
+            cb_->socket().ioHandle().resetFileEvents();
+            // Do not skip following listener filters.
+            cb_->continueFilterChain(true);
+            break;
+          case ParseState::Continue:
+            // do nothing but wait for the next event
+            break;
+          }
+        },
+        Event::PlatformDefaultTriggerType, Event::FileReadyType::Read);
```

Comment:
BTW, should we handle a peer closing connection?

Reply:
The original listener filter code did register both read and closed events, but closed event is removed due to an issue on Windows, please see https://github.com/envoyproxy/envoy/pull/18265